### PR TITLE
feat: add support for allowDestructiveSchemaUpdates in defineData

### PR DESCRIPTION
**Pull request by alharris-at**
*Issue #, if available:*
N/A

*Description of changes:*
When AMPLIFY_TABLE deployment strategy was enabled, we didn't add the requisite override flag to enable deploying destructive updates (e.g. key changes). This PR wires that through, and adds test for default, positive, and negative cases.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/backend-data/src/types.ts
Comment (palpatim) [inline packages/backend-data/src/types.ts:138]:
```suggestion
   * It is recommended to only enable this if you plan on making a destructive change, and disable it again in a subsequent deployment.
```

**Review by palpatim — COMMENTED** [review]
minor wording nit

**Review by AaronZyLee — APPROVED** [review]

File: packages/backend-data/src/types.ts
Comment (edwardfoyle) [inline packages/backend-data/src/types.ts:141]: I'm not sure we should expose this at this level. My vote would be to take a similar approach to what we did with auth UserPools and _always_ delete. We will catch potentially destructive behavior at the CI/CD level in hosting by looking at CFN ChangeSets before doing the deployment.

cc @swaminator @srquinn21

Comment (alharris-at) [reply packages/backend-data/src/types.ts:141]: This is a different use-case. Customer tables will always be dropped on stack deletion, this enables the ability to push pk changes to your schema without first removing the model from your schema. Consistent with --allow-destructive-updates in Gen1

Comment (edwardfoyle) [reply packages/backend-data/src/types.ts:141]: The use case is still "making an update that will cause deletion/recreation of a stateful resource. The stance we're taking is that it should "just do it" and deletion protections should be an upstream concern of the CICD checks

Comment (alharris-at) [reply packages/backend-data/src/types.ts:141]: I think the nuance of "things that can cause data loss" while updating the schema is too fine for me to by comfortable just allowing destructive actions by default. I'd rather not ship the ability to enable this and force a user to 1/ push a dropped table, then 2/ push again with the new table definition than allow them to accidentally push an update that will delete data if it was unintentional.

Comment (sobolk) [reply packages/backend-data/src/types.ts:141]: My 0.02.

1. Whatever we do with destructive updates, the solution should be consistent across auth data and future verticals.
    1. Default should be same. But can be different sandbox vs pipeline deployment.
    2. Perhaps there should be a high level knob that's honored by any vertical.
2. Imo:
    1. Sandbox should destroy by default
    2. Pipeline should retain by default
    3. There should be ephemeral knob in hosting console to roll out breaks if that's desired

Comment (alharris-at) [reply packages/backend-data/src/types.ts:141]: Desire for consistency makes sense, and long-term I'm not opposed to console exposing the 'hook' or action to have a human intervene and approve here if necessary. I'll chat w/ some folks about the feasibility of getting those lined up tomorrow, then either close or update this coming out of those chats. Feedback appreciated!

Comment (sobolk) [reply packages/backend-data/src/types.ts:141]: Do the checks happen at synth or deploy attempt ?

Comment (alharris-at) [reply packages/backend-data/src/types.ts:141]: The particular checks we're talking about here are executed during the cloudformation deployment in a custom resource.

**Comment by alharris-at** [comment]
Closing this, and likely converting to a flag that comes in the `pipeline-deploy` command.

**Pull request closed without merging**
